src/main/kotlin/org/rust/ide/intentions/DemorgansLawIntention.kt

```diff
@@ -0,0 +1,138 @@
+package org.rust.ide.intentions
+
+import com.intellij.codeInsight.intention.PsiElementBaseIntentionAction
+import com.intellij.openapi.editor.Editor
+import com.intellij.openapi.project.Project
+import com.intellij.psi.PsiElement
+import com.intellij.psi.tree.IElementType
+import org.rust.lang.core.psi.*
+import org.rust.lang.core.psi.util.parentOfType
+
+class DemorgansLawIntention : PsiElementBaseIntentionAction() {
+    private var m_text = ""
+    override fun getText() = m_text
+    override fun getFamilyName() = text
+    override fun startInWriteAction() = true
+
+    fun getTextForElement(element: PsiElement) {
+        val binaryExpression = element as RustBinaryExprElement
+        m_text = when (binaryExpression.operatorType) {
+            RustTokenElementTypes.ANDAND -> "DeMorgan's Law, Replace '&&' with '||'"
+            RustTokenElementTypes.OROR -> "DeMorgan's Law, Replace '||' with '&&'"
+            else -> ""
+        }
+    }
+
+    private data class Context(
+        val binaryExpr: RustBinaryExprElement,
+        val binaryOpType: IElementType
+    )
+
+    private fun findContext(element: PsiElement): Context? {
+        if (!element.isWritable) return null
+
+        val binExpr = element.parentOfType<RustBinaryExprElement>() ?: return null
+        val opType = binExpr.operatorType
+        if (opType == RustTokenElementTypes.ANDAND || opType == RustTokenElementTypes.OROR) {
+            return Context(binExpr, opType)
+        }
+        return null
+    }
+
+    override fun invoke(project: Project, editor: Editor?, element: PsiElement) {
+        val (binExpr, opType) = findContext(element) ?: return
+
+        var topBinExpr = binExpr
+        while (topBinExpr.parent is RustBinaryExprElement && (topBinExpr.parent as RustBinaryExprElement).operatorType == opType) {
+            topBinExpr = topBinExpr.parent as RustBinaryExprElement
+        }
+
+        val converted = convertConjunctionExpression(topBinExpr, opType) ?: return
+
+        var expressionToReplace: RustExprElement = topBinExpr
+        var expString = "!($converted)"
+        val parent = topBinExpr.parent.parent
+        if (isNegation(parent)) {
+            expressionToReplace = parent as RustExprElement
+            expString = converted
+        }
+        val newExpr = RustElementFactory.createExpression(project, expString) ?: return
+
+        expressionToReplace.replace(newExpr)
+    }
+
+    override fun isAvailable(project: Project, editor: Editor?, element: PsiElement): Boolean {
+        if (findContext(element) != null) {
+            getTextForElement(findContext(element)?.binaryExpr as PsiElement)
+            return true
+        }
+        return false
+    }
+
+    private fun isConjunctionExpression(expression: RustExprElement, opType: IElementType): Boolean {
+        return expression is RustBinaryExprElement && expression.operatorType == opType
+    }
+
+    private fun getNegated(expression: PsiElement): RustExprElement? {
```

Comment:
It's better to inline this function because it is used only once and it is not obvious what it does without the preceding check.